### PR TITLE
feat(validation): sequencer misbehavior detection 

### PR DESCRIPTION
**Pull request by faultytolly**
# PR Standards

## Opening a pull request should be able to meet the following requirements

--

PR naming convention: https://hackmd.io/@nZpxHZ0CT7O5ngTp0TP9mg/HJP_jrm7A

---

Close #XXX

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: cambios.patch
Comment (mtsitrin) [inline cambios.patch:1]: what's this one?

File: block/slvalidator.go [hunk 1]
```diff
@@ -170,6 +170,31 @@ func (v *SettlementValidator) ValidateDaBlocks(slBatch *settlement.ResultRetriev
...
 			return err
 		}
+
+		// we compare the sequencer address between SL state info and DA block
+		// if next sequencer is not set, we check if the sequencer hash is equal to the next sequencer hash
+		// because it did not change. If the next sequencer is set, we check if the next sequencer hash is equal on the
+		// last block of the batch
+		isLastBlock := i == len(slBatch.BlockDescriptors)-1
+		if slBatch.NextSequencer != "" && isLastBlock {
+			nextSequencer := v.blockManager.State.Sequencers.GetByAddress(slBatch.NextSequencer)
```
Comment (omritoptix) [inline block/slvalidator.go:180]: I blv this api is deprecated now after @keruch change as we don't directly save the `Sequencers` on the state

Comment (omritoptix) [inline block/slvalidator.go:180]: what if we don't have the nextSequencer in the store? we should probably fetch all the sequencers again (we can't count on polling to be exact in terms of syncing). we can use https://github.com/dymensionxyz/dymint/pull/1163/files#diff-c5773c2f4ace59a21cd36af8a0c95b46135151ea34f2b4bc09e6cbecc0c3e46aR169

File: block/slvalidator.go [hunk 2]
```diff
@@ -170,6 +170,31 @@ func (v *SettlementValidator) ValidateDaBlocks(slBatch *settlement.ResultRetriev
...
+			var nextSequencerHash [32]byte
+			copy(nextSequencerHash[:], nextSequencer.MustHash())
+
+			if !bytes.Equal(nextSequencer.MustHash(), daBlocks[i].Header.NextSequencersHash[:]) {
+				return types.NewErrInvalidNextSequencersHashFraud(
+					nextSequencerHash,
+					daBlocks[i].Header.NextSequencersHash,
+				)
+			}
```
Comment (omritoptix) [inline block/slvalidator.go:189]: This var seems redundant
```go
var nextSequencerHash [32]byte
```
possible suggestion (Wasn't tested and probably identation errors): 
```suggestion
            nextSequencerHash := nextSequencer.MustHash()
            if !bytes.Equal(nextSequencerHash, daBlocks[i].Header.NextSequencersHash[:]) {
                return types.NewErrInvalidNextSequencersHashFraud(
                    *(*[32]byte)(nextSequencerHash),
                    daBlocks[i].Header.NextSequencersHash,
                )
            }
```

File: settlement/dymension/events.go [hunk 3]
```diff
@@ -111,26 +111,35 @@ func convertToNewBatchEvent(rawEventData ctypes.ResultEvent) (*settlement.EventD
...
 		return nil, errors.Join(errs...)
 	}
 	endHeight := uint64(startHeight + numBlocks - 1)
+
+	var nextProposer string
+	if events["state_update.next_proposer"] == nil {
+		nextProposer = ""
+	} else {
+		nextProposer = events["state_update.next_proposer"][0]
+	}
```
Comment (omritoptix) [inline settlement/dymension/events.go:136]: what is this part used for?
if im not mistaken we only use the new batch to get informed about a new batch and than fetch the full batch from the hub for syncing by it's height (as other fields like block descriptors etc are also absent)

**Review by omritoptix — CHANGES_REQUESTED** [review]

File: coverage.txt
Comment (mtsitrin) [inline coverage.txt:1]: mistake?

File: block/slvalidator.go [hunk 4]
```diff
@@ -146,6 +146,11 @@ func (v *SettlementValidator) ValidateDaBlocks(slBatch *settlement.ResultRetriev
 		return types.NewErrStateUpdateNumBlocksNotMatchingFraud(slBatch.EndHeight, numSLBlocks, numSLBlocks)
 	}
 
+	currentProposer := v.blockManager.State.GetProposer()
+	if currentProposer == nil {
```
Comment (mtsitrin) [inline block/slvalidator.go:150]: will it allow historical sync of a halted rollapp? (rollapp without active proposer)

Comment (omritoptix) [reply block/slvalidator.go:150]: as we're getting this from the state, and the state should reflect the rollapp proposer for height (it's not currently but there is an open issue on it), it means that by the time we've got here the chain is most likely halted (i.e has no current proposer).

in that case we have nothing to validate.

**Review by omritoptix — APPROVED** [review]

**Pull request merged**